### PR TITLE
feat: addon disable and uninstall check

### LINKED ISSUE #192: [Improvement] check if there are clusters using addons when disable/uninstall addons
When disable or uninstall addons, check if there are any clusters currently using this addon. If there are, please inform the user that these clusters may be affected.

And add double-check.

### PR DESCRIPTION
**Pull request by 1aal**
- fix #192

image

### CODE REVIEW

File: pkg/cmd/addon/addon.go [hunk 1]
```diff
@@ -305,6 +309,7 @@ func (o *addonCmdOpts) fetchAddonObj() error {
 }
 
 func (o *addonCmdOpts) validate() error {
+
```
Comment (ldming) [inline pkg/cmd/addon/addon.go:312]: revert this line

File: pkg/cmd/addon/util.go [hunk 2]
```diff
@@ -40,3 +49,33 @@ func getAddonVersion(addon *extensionsv1alpha1.Addon) string {
 	}
 	return ""
 }
+
+func CheckBeforeDisableAddon(f cmdutil.Factory, addons []string) error {
```
Comment (ldming) [inline pkg/cmd/addon/util.go:53]: This function is used by `disable` and `uninstall`, so its name is not appropriate.

Maybe `CheckAddonUsedByCluster`

File: pkg/cmd/addon/util.go [hunk 3]
```diff
@@ -40,3 +49,33 @@ func getAddonVersion(addon *extensionsv1alpha1.Addon) string {
...
+	} else if len(infos) != 0 {
+		errMsg := "There are addons are being used:\n"
+		for _, info := range infos {
+			var cluster v1alpha1.Cluster
+			if err = runtime.DefaultUnstructuredConverter.FromUnstructured(info.Object.(*unstructured.Unstructured).Object, &cluster); err != nil {
+				return err
+			}
+			errMsg += fmt.Sprintf("name: %s namespace: %s addon: %s\n", printer.BoldRed(info.Name), printer.BoldYellow(info.Namespace), printer.BoldRed(cluster.Labels[constant.ClusterDefLabelKey]))
+		}
+		errMsg += "please delete the cluster(s) first!\n"
```
Comment (ldming) [inline pkg/cmd/addon/util.go:76]: If there are clusters using these addons, notify user these clusters and let user confirm if disable/uninstall.

File: pkg/cmd/addon/util.go [hunk 4]
```diff
@@ -40,3 +49,33 @@ func getAddonVersion(addon *extensionsv1alpha1.Addon) string {
...
+	labelSelecotor := util.BuildClusterLabel("", addons)
+	r := f.NewBuilder().
+		Unstructured().
+		AllNamespaces(true).
+		LabelSelector(labelSelecotor).
+		ResourceTypeOrNameArgs(true, []string{util.GVRToString(types.ClusterGVR())}...).
+		ContinueOnError().
+		Latest().
+		Flatten().
+		Do()
```
Comment (ldming) [inline pkg/cmd/addon/util.go:63]: I prefer to build a dynamic client by cmdutil.Factory and List the clusters, feels more intuitive to me personally.

**Comment by 1aal** [comment]
- Bad branch name.
- reopen in https://github.com/apecloud/kbcli/pull/202

**Pull request closed without merging**